File: src/main/java/com/polytech/webscraipper/sdk/Prompts.java
package com.polytech.webscraipper.sdk;

import com.polytech.webscraipper.sdk.responses.PromptResponse;
import com.polytech.webscraipper.sdk.responses.PromptsResponse;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;

public class Prompts {

  private final PromptsClient promptsClient;

  public Prompts(PromptsClient promptsClient) {
    this.promptsClient = promptsClient;
  }

  public List<String> getPromptVariables(String promptName, Integer version, String label) {
    PromptResponse prompt = getByName(promptName, version, label);
    return extractVariablesFromText(prompt.prompt);
  }

  public PromptResponse getCustomizedPrompt(
      String promptName, Integer version, String label, Map<String, String> variables) {
    PromptResponse prompt = getByName(promptName, version, label);
    prompt.prompt = replacesVariablesInText(prompt.prompt, variables);
    return prompt;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(
      String name,
      String label,
      String tag,
      Integer page,
      Integer limit,
      String fromUpdatedAt,
      String toUpdatedAt) {
    return promptsClient.getAllPrompts(name, label, tag, page, limit, fromUpdatedAt, toUpdatedAt)
        .data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(
      String name, String label, String tag, Integer page, Integer limit, String fromUpdatedAt) {
    return promptsClient.getAllPrompts(name, label, tag, page, limit, fromUpdatedAt, null).data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(
      String name, String label, String tag, Integer page, Integer limit) {
    return promptsClient.getAllPrompts(name, label, tag, page, limit, null, null).data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(
      String name, String label, String tag, Integer page) {
    return promptsClient.getAllPrompts(name, label, tag, page, null, null, null).data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(
      String name, String label, String tag) {
    return promptsClient.getAllPrompts(name, label, tag, null, null, null, null).data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(String name, String label) {
    return promptsClient.getAllPrompts(name, label, null, null, null, null, null).data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts(String name) {
    return promptsClient.getAllPrompts(name, null, null, null, null, null, null).data;
  }

  public List<PromptsResponse.IndividualPrompt> getAllPrompts() {
    return promptsClient.getAllPrompts(null, null, null, null, null, null, null).data;
  }

  public PromptResponse getByName(String promptName, Integer version, String label) {
    return promptsClient.getPromptByName(promptName, version, label);
  }

  @FeignClient(
      name = "langfuse-sdk-prompts",
      url = "https://cloud.langfuse.com/api/public/v2/prompts",
      configuration = FeignConfig.class)
  public interface PromptsClient {

    @GetMapping("/")
    PromptsResponse getAllPrompts(
        @RequestParam(value = "name", required = false) String name,
        @RequestParam(value = "label", required = false) String label,
        @RequestParam(value = "tag", required = false) String tag,
        @RequestParam(value = "page", required = false) Integer page,
        @RequestParam(value = "limit", required = false) Integer limit,
        @RequestParam(value = "fromUpdatedAt", required = false) String fromUpdatedAt,
        @RequestParam(value = "toUpdatedAt", required = false) String toUpdatedAt);

    @GetMapping("/{promptId}")
    PromptResponse getPromptByName(
        @PathVariable("promptId") String promptId,
        @RequestParam(value = "version", required = false) Integer version,
        @RequestParam(value = "label", required = false) String label);
  }

  public List<String> extractVariablesFromText(String text) {
    Pattern pattern = Pattern.compile("\\{\\{([^}]*)}}");
    Matcher matcher = pattern.matcher(text);
    List<String> variables = new ArrayList<>();
    while (matcher.find()) {
      variables.add(matcher.group(1));
    }
    return variables;
  }

  public String replacesVariablesInText(String text, Map<String, String> variables) {
    List<String> requiredVariables = extractVariablesFromText(text);
    if (!variables.keySet().containsAll(requiredVariables)) {
      throw new IllegalArgumentException("Not all variables have a value");
    }
    for (Map.Entry<String, String> entry : variables.entrySet()) {
      text = text.replace("{{" + entry.getKey() + "}}", entry.getValue());
    }
    return text;
  }
}

File: src/main/java/com/polytech/webscraipper/sdk/Prompts.kt
package com.polytech.webscraipper.sdk

import com.polytech.webscraipper.sdk.responses.PromptResponse
import com.polytech.webscraipper.sdk.responses.PromptsResponse
import org.springframework.cloud.openfeign.FeignClient
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam

class Prompts(private val promptsClient: PromptsClient) {

    @JvmOverloads
    fun getPromptVariables(promptName: String, version: Int? = null, label: String? = null): List<String> {
        val prompt = getByName(promptName, version, label)
        return extractVariablesFromText(prompt.prompt)
    }

    @JvmOverloads
    fun getCustomizedPrompt(promptName: String, version: Int? = null, label: String? = null, variables: Map<String, String>): Pair<String, PromptResponse> {
        val prompt = getByName(promptName, version, label)
        return replacesVariablesInText(prompt.prompt, variables) to prompt
    }

    @JvmOverloads
    fun getAllPrompts(
        name: String? = null,
        label: String? = null,
        tag: String? = null,
        page: Int? = null,
        limit: Int? = null,
        fromUpdatedAt: String? = null,
        toUpdatedAt: String? = null,
    ): List<PromptsResponse.IndividualPrompt> = promptsClient.getAllPrompts(name, label, tag, page, limit, fromUpdatedAt, toUpdatedAt).data

    @JvmOverloads
    fun getByName(promptName: String, version: Int? = null, label: String? = null): PromptResponse = promptsClient.getPromptByName(promptName, version, label)

    @FeignClient(
        name = "langfuse-sdk-prompts",
        url = "https://cloud.langfuse.com/api/public/v2/prompts",
        configuration = [FeignConfig::class],
    )
    interface PromptsClient {

        @GetMapping("/")
        fun getAllPrompts(
//            name
//            null | string
//        label
//        null | string
//        tag
//        null | string
//        page
//        null | integer
//        page number, starts at 1
//
//        limit
//        null | integer
//        limit of items per page
//
//        fromUpdatedAt
//        null | string
//        date-time
//        Optional filter to only include prompt versions created/updated on or after a certain datetime (ISO 8601)
//
//        toUpdatedAt
//        null | string
//        date-time
//        Optional filter to only include prompt versions created/updated before a certain datetime (ISO 8601)
            @RequestParam("name", required = false) name: String? = null,
            @RequestParam("label", required = false) label: String? = null,
            @RequestParam("tag", required = false) tag: String? = null,
            @RequestParam("page", required = false) page: Int? = null,
            @RequestParam("limit", required = false) limit: Int? = null,
            @RequestParam("fromUpdatedAt", required = false) fromUpdatedAt: String? = null,
            @RequestParam("toUpdatedAt", required = false) toUpdatedAt: String? = null,
        ): PromptsResponse

        @GetMapping("/{promptId}")
        fun getPromptByName(
            @PathVariable("promptId") promptId: String,
            @RequestParam("version", required = false) version: Int? = null,
            @RequestParam("label", required = false) label: String? = null,
        ): PromptResponse
    }

    fun extractVariablesFromText(text: String): List<String> {
        val regex = Regex("\\{\\{([^}]*)}}")
        return regex.findAll(text).map { it.groupValues[1] }.toList()
    }

    fun replacesVariablesInText(text: String, variables: Map<String, String>): String {
        var newText = text

        val reelVariables = extractVariablesFromText(text)

        if (!variables.keys.containsAll(reelVariables)) {
            throw IllegalArgumentException("Not all variables have a value")
        }

        variables.forEach { (key, value) -> newText = newText.replace("{{$key}}", value) }
        return newText
    }
}
